File: Kiters/Kiters Watch App/Services/WatchSessionUploader.swift
import Foundation

// Response types
struct StartResponse {
    let sessId: Int
    let spot: String
    let poiKind: String
}

struct RecordResponse {
    let broken: [String]
}

struct EndResponse {
    let broken: [String]
}

enum UploaderError: LocalizedError {
    case notAuthenticated
    case networkError(String)
    case serverError(Int, String)

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:          return "Not signed in — please log in again."
        case .networkError(let msg):     return "Network error: \(msg)"
        case .serverError(let s, let m): return "Server error \(s): \(m)"
        }
    }
}

/// HTTP client for the four watch-ingest lifecycle calls.
/// Every method fetches the current JWT from WatchPairingStore (auto-refreshing
/// if needed) and POSTs to watch-ingest with the correct body shape.
final class WatchSessionUploader {
    static let shared = WatchSessionUploader()

    private let ingestURL = URL(string: "https://vvowvcdylztsqpzifdqc.supabase.co/functions/v1/watch-ingest")!
    private let anonKey   = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InZ2b3d2Y2R5bHp0c3FwemlmZHFjIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzU0MTc1NDcsImV4cCI6MjA5MDk5MzU0N30.jPBYr6f9fTABLHAD1rY_b1HP8xI0cDEQPJczxjCKsSY"

    // MARK: - start

    /// Opens a live session. Call once after the first GPS fix.
    /// Returns sessId (keep it for all subsequent calls), spot name, and POI kind.
    func start(lat: Double, lng: Double, startedAt: Date = Date()) async throws -> StartResponse {
        let iso = ISO8601DateFormatter().string(from: startedAt)
        let body: [String: Any] = ["type": "start", "lat": lat, "lng": lng, "startedAt": iso]
        let json = try await post(body)
        guard let sessId = json["sessId"] as? Int else {
            throw UploaderError.networkError("Missing sessId in start response")
        }
        return StartResponse(
            sessId:  sessId,
            spot:    json["spot"]    as? String ?? "",
            poiKind: json["poiKind"] as? String ?? ""
        )
    }

    // MARK: - ping

    /// Position heartbeat — call every ~10 s while riding.
    func ping(sessId: Int, lat: Double, lng: Double, jmax: Double? = nil, jcnt: Int? = nil) async throws {
        var body: [String: Any] = ["type": "ping", "sessId": sessId, "lat": lat, "lng": lng]
        if let jmax { body["jmax"] = jmax }
        if let jcnt { body["jcnt"] = jcnt }
        _ = try await post(body)
    }

    // MARK: - record

    /// Call when a session-best metric improves. Pass only the metric(s) that improved.
    /// The server decides if it also beats the all-time personal best.
    @discardableResult
    func record(sessId: Int, jumpM: Double? = nil, airS: Double? = nil,
                speedKmh: Double? = nil, distKm: Double? = nil) async throws -> RecordResponse {
        var body: [String: Any] = ["type": "record", "sessId": sessId]
        if let jumpM    { body["jumpM"]    = jumpM    }
        if let airS     { body["airS"]     = airS     }
        if let speedKmh { body["speedKmh"] = speedKmh }
        if let distKm   { body["distKm"]   = distKm   }
        let json = try await post(body)
        return RecordResponse(broken: json["broken"] as? [String] ?? [])
    }

    // MARK: - end

    /// Finalises the session. Provide full metrics, compact jData, and track.
    @discardableResult
    func end(sessId: Int,
             durMin: Int, jmax: Double, jcnt: Int, airS: Double,
             spdKmh: Int, distKm: Double,
             windKts: Int? = nil, dir: String? = nil, avgKmh: Double? = nil,
             stars: Int = 3,
             track: [[Int]], jData: [[String: Int]]) async throws -> EndResponse {
        var body: [String: Any] = [
            "type":   "end",
            "sessId": sessId,
            "durMin": durMin,
            "jmax":   jmax,
            "jcnt":   jcnt,
            "airS":   airS,
            "spdKmh": spdKmh,
            "distKm": distKm,
            "stars":  stars,
            "track":  track,
            "jData":  jData,
        ]
        if let windKts { body["windKts"] = windKts }
        if let dir     { body["dir"]     = dir     }
        if let avgKmh  { body["avgKmh"]  = avgKmh  }
        let json = try await post(body)
        return EndResponse(broken: json["broken"] as? [String] ?? [])
    }

    // MARK: - Private

    private func post(_ body: [String: Any]) async throws -> [String: Any] {
        guard let pairing = try? await WatchPairingStore.shared.validPairing() else {
            throw UploaderError.notAuthenticated
        }
        var req = URLRequest(url: ingestURL)
        req.httpMethod = "POST"
        req.setValue("application/json",              forHTTPHeaderField: "Content-Type")
        req.setValue(anonKey,                         forHTTPHeaderField: "apikey")
        req.setValue("Bearer \(pairing.accessToken)", forHTTPHeaderField: "Authorization")
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        let (data, resp) = try await URLSession.shared.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0

        if status == 401 {
            await WatchPairingStore.shared.clear()
            throw UploaderError.notAuthenticated
        }
        guard (200..<300).contains(status) else {
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String ?? ""
            throw UploaderError.serverError(status, msg)
        }
        return (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
    }
}

File: android/wear/src/main/java/com/spoteq/wear/session/WatchSessionUploader.kt
package com.spoteq.wear.session

import com.spoteq.wear.auth.AuthRepository
import com.spoteq.wear.auth.AuthSession
import com.spoteq.wear.storage.SettingsStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

data class StartResponse(val sessId: Int, val spot: String, val poiKind: String)
data class RecordResponse(val broken: List<String>)
data class EndResponse(val broken: List<String>)

sealed class UploaderError(message: String) : Exception(message) {
    object NotAuthenticated : UploaderError("Not signed in — please log in again.")
    data class ServerError(val status: Int, val body: String) : UploaderError("Server error $status: $body")
    data class NetworkError(override val message: String) : UploaderError(message)
}

/**
 * HTTP client for the four watch-ingest lifecycle calls.
 * Reads the current JWT from SettingsStore, refreshing via AuthRepository if needed.
 */
class WatchSessionUploader(
    private val settings: SettingsStore,
    private val authRepo: AuthRepository = AuthRepository(),
) {
    private val ingestUrl = "https://vvowvcdylztsqpzifdqc.supabase.co/functions/v1/watch-ingest"
    private val anonKey   = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InZ2b3d2Y2R5bHp0c3FwemlmZHFjIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzU0MTc1NDcsImV4cCI6MjA5MDk5MzU0N30.jPBYr6f9fTABLHAD1rY_b1HP8xI0cDEQPJczxjCKsSY"

    // MARK: - start

    suspend fun start(lat: Double, lng: Double, startedAt: Date = Date()): Result<StartResponse> =
        withContext(Dispatchers.IO) {
            runCatching {
                val iso = isoDate(startedAt)
                val body = JSONObject().apply {
                    put("type", "start"); put("lat", lat); put("lng", lng); put("startedAt", iso)
                }
                val json = post(body)
                StartResponse(
                    sessId  = json.getInt("sessId"),
                    spot    = json.optString("spot", ""),
                    poiKind = json.optString("poiKind", ""),
                )
            }
        }

    // MARK: - ping

    suspend fun ping(sessId: Int, lat: Double, lng: Double,
                     jmax: Double? = null, jcnt: Int? = null): Result<Unit> =
        withContext(Dispatchers.IO) {
            runCatching {
                val body = JSONObject().apply {
                    put("type", "ping"); put("sessId", sessId); put("lat", lat); put("lng", lng)
                    jmax?.let { put("jmax", it) }
                    jcnt?.let { put("jcnt", it) }
                }
                post(body)
                Unit
            }
        }

    // MARK: - record

    suspend fun record(sessId: Int, jumpM: Double? = null, airS: Double? = null,
                       speedKmh: Double? = null, distKm: Double? = null): Result<RecordResponse> =
        withContext(Dispatchers.IO) {
            runCatching {
                val body = JSONObject().apply {
                    put("type", "record"); put("sessId", sessId)
                    jumpM?.let    { put("jumpM", it)    }
                    airS?.let     { put("airS", it)     }
                    speedKmh?.let { put("speedKmh", it) }
                    distKm?.let   { put("distKm", it)   }
                }
                val json = post(body)
                RecordResponse(broken = jsonArrayToList(json.optJSONArray("broken")))
            }
        }

    // MARK: - end

    suspend fun end(
        sessId: Int,
        durMin: Int, jmax: Double, jcnt: Int, airS: Double,
        spdKmh: Int, distKm: Double,
        windKts: Int? = null, dir: String? = null, avgKmh: Double? = null,
        stars: Int = 3,
        track: List<List<Int>>,
        jData: List<Map<String, Int>>,
    ): Result<EndResponse> = withContext(Dispatchers.IO) {
        runCatching {
            val trackArr = JSONArray().apply { track.forEach { pt -> add(JSONArray().apply { pt.forEach { add(it) } }) } }
            val jDataArr = JSONArray().apply { jData.forEach { j -> add(JSONObject(j as Map<*, *>)) } }
            val body = JSONObject().apply {
                put("type", "end"); put("sessId", sessId)
                put("durMin", durMin); put("jmax", jmax); put("jcnt", jcnt)
                put("airS", airS); put("spdKmh", spdKmh); put("distKm", distKm)
                put("stars", stars); put("track", trackArr); put("jData", jDataArr)
                windKts?.let { put("windKts", it) }
                dir?.let     { put("dir", it)     }
                avgKmh?.let  { put("avgKmh", it)  }
            }
            val json = post(body)
            EndResponse(broken = jsonArrayToList(json.optJSONArray("broken")))
        }
    }

    // MARK: - Private

    private suspend fun accessToken(): String {
        val token    = settings.authAccessToken
        val expires  = settings.authExpiresAt
        val refresh  = settings.authRefreshToken
        if (token.isBlank()) throw UploaderError.NotAuthenticated
        val nowSec = System.currentTimeMillis() / 1000L
        if (expires - nowSec < 120 && refresh.isNotBlank()) {
            authRepo.refreshToken(refresh).getOrNull()?.let { session: AuthSession ->
                settings.authAccessToken  = session.accessToken
                settings.authRefreshToken = session.refreshToken
                settings.authExpiresAt    = session.expiresAt
                return session.accessToken
            }
        }
        return token
    }

    private suspend fun post(body: JSONObject): JSONObject {
        val jwt = accessToken()
        val conn = (URL(ingestUrl).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type",  "application/octet-stream")
            setRequestProperty("apikey",        anonKey)
            setRequestProperty("Authorization", "Bearer $jwt")
            doOutput     = true
            connectTimeout = 15_000
            readTimeout    = 15_000
        }
        conn.outputStream.use { os ->
            os.write(BinaryLogEnvelope.encode(body))
        }
        val status = conn.responseCode
        if (status == 401) {
            settings.authAccessToken = ""
            conn.disconnect()
            throw UploaderError.NotAuthenticated
        }
        val responseBody = if (status in 200..299)
            conn.inputStream.bufferedReader().readText()
        else
            conn.errorStream?.bufferedReader()?.readText() ?: ""
        conn.disconnect()
        if (status !in 200..299) throw UploaderError.ServerError(status, responseBody)
        return if (responseBody.isNotBlank()) JSONObject(responseBody) else JSONObject()
    }

    private fun jsonArrayToList(arr: JSONArray?): List<String> {
        if (arr == null) return emptyList()
        return (0 until arr.length()).map { arr.getString(it) }
    }

    private fun isoDate(date: Date): String {
        val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        fmt.timeZone = TimeZone.getTimeZone("UTC")
        return fmt.format(date)
    }

    private fun JSONArray.add(value: Any) = put(value)
}
